src/ripple/consensus/LedgerTrie.h

```diff
@@ -0,0 +1,712 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2017 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_APP_CONSENSUS_LEDGERS_TRIE_H_INCLUDED
+#define RIPPLE_APP_CONSENSUS_LEDGERS_TRIE_H_INCLUDED
+
+#include <algorithm>
+#include <memory>
+#include <vector>
+#include <ripple/json/json_value.h>
+
+namespace ripple {
+
+/** Ancestry trie of ledgers
+
+    Combination of a compressed trie and merkle-ish tree that maintains
+    validation support of recent ledgers based on their ancestry.
+
+    The compressed trie structure comes from recognizing that ledger history
+    can be viewed as a string over the alphabet of ledger ids. That is,
+    a given ledger with sequence number `seq` defines a length `seq` string,
+    with i-th entry equal to the id of the ancestor ledger with sequence
+    number i. "Sequence" strings with a common prefix share those ancestor
+    ledgers in common. Tracking this ancestry information and relations across
+    all validated ledgers is done conveniently in a compressed trie. A node in
+    the trie is an ancestor of all its children. If a parent node has sequence
+    number `seq`, each child node has a different ledger starting at `seq+1`.
+    The compression comes from the invariant that any non-root node with 0 tip
+    support has either no children or multiple children. In other words, a
+    non-root 0-tip-support node can be combined with its single child.
+
+    The merkle-ish property is based on the branch support calculation. Each
+    node has a tipSupport, which is the number of current validations for that
+    particular ledger. The branch support is the sum of the tip support and
+    the branch support of that node's children:
+
+        @code
+        node->branchSupport = node->tipSupport
+                            + sum_(child : node->children) child->branchSupport
+        @endcode
+
+    This is analagous to the merkle tree property in which a node's hash is
```

Comment:
Nit: Found some misspellings (analagous -> analogous). Here's a commit with fixes for the ones I found: https://github.com/seelabs/rippled/commit/5a3e182bbc44003f7f0fb8bad1d3401647d1c70c